google-cloud-firestore/src/main/java/com/google/cloud/firestore/Watch.java

```diff
@@ -0,0 +1,580 @@
+   * stream.
+   */
+  private static final int WATCH_TARGET_ID = 0xD0;
+
+  private final FirestoreImpl firestore;
+  private final Comparator<DocumentSnapshot> comparator;
+  private final ExponentialBackoff backoff;
+  private final Target target;
+  private ApiStreamObserver<ListenRequest> stream;
+
+  /** The sorted tree of DocumentSnapshots as sent in the last snapshot. */
+  private DocumentSet documentSet;
+
+  /** The accumulates map of document changes (keyed by document name) for the current snapshot. */
+  private Map<ResourcePath, Document> changeMap;
+
+  /** The server-assigned resume token. */
+  private ByteString resumeToken;
+
+  /** The user-provided listener. */
+  private EventListener<QuerySnapshot> listener;
+
+  /** The user-provided callback executor. */
+  private Executor userCallbackExecutor;
+
+  /**
+   * Whether the retrieved result set has been marked 'CURRENT' (has caught up with the backend
+   * state at target initialization).
+   */
+  private boolean current;
+
+  /**
+   * Tracks whether we've pushed an initial set of changes. This is needed since we should push
+   * those even when there are none.
+   */
+  private boolean hasPushed;
+
+  /**
+   * Indicates whether we are interested in data from the stream. Set to false
+   * in the 'unsubscribe()' callback.
+   */
+  private AtomicBoolean isActive;
+
+  /** The list of document changes in a snapshot separated by change type. */
+  static class ChangeSet {
+    List<DocumentSnapshot> deletes = new ArrayList<>();
+    List<DocumentSnapshot> adds = new ArrayList<>();
+    List<DocumentSnapshot> updates = new ArrayList<>();
+  }
+
+  /**
+   * @param firestore The Firestore Database client.
+   * @param target A Firestore 'Target' proto denoting the target to listen on.
+   * @param comparator A comparator for DocumentSnapshots that is used to order the document
+   *     snapshots returned by this watch.
+   */
+  private Watch(FirestoreImpl firestore, Target target, Comparator<DocumentSnapshot> comparator) {
+    this.firestore = firestore;
+    this.target = target;
+    this.comparator = comparator;
+    this.backoff = new ExponentialBackoff(firestore.getClient().getExecutor());
+    this.isActive = new AtomicBoolean();
+  }
+
+  /**
+   * Creates a new Watch instance that listens on DocumentReferences.
+   *
+   * @param documentReference The document reference for this watch.
+   * @return A newly created Watch instance.
+   */
+  static Watch forDocument(DocumentReference documentReference) {
+    Target.Builder target = Target.newBuilder();
+    target.getDocumentsBuilder().addDocuments(documentReference.getName());
+    target.setTargetId(WATCH_TARGET_ID);
+
+    return new Watch(
+        (FirestoreImpl) documentReference.getFirestore(),
+        target.build(),
+        new Comparator<DocumentSnapshot>() {
+          @Override
+          public int compare(DocumentSnapshot o1, DocumentSnapshot o2) {
+            return 0;
```

Comment:
I don't really understand why. Is it because we'll only ever compare a DocumentSnapshot against itself?

Reply:
This should actually never need to get called because the number of documents in a DocumentSnapshot is always just one. I added an assert and a comment.